File: utils/src/main/java/ru/virgil/utils/fluent_request/BodyStepMultipart.java
package ru.virgil.utils.fluent_request;

import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.mock.web.MockMultipartFile;

import javax.annotation.Nullable;
import java.util.List;

@RequiredArgsConstructor
public class BodyStepMultipart implements BodyStepMultipartStart, BodyStep {

    private final RequestModel requestModel;
    private final Requester requester;
    private final ObjectMapper objectMapper;
    @Nullable
    private JavaType byteArrayJavaType;

    @Override
    public BodyStepMultipart file(MockMultipartFile mockMultipartFile) throws Exception {
        requestModel.setMockMultipartFile(mockMultipartFile);
        return this;
    }

    public BodyStepMultipart receive(Class<?> responseClass, Class<?>... responseClasses) {
        JavaType javaType = objectMapper.getTypeFactory().constructParametricType(responseClass, responseClasses);
        requestModel.setResponseJavaType(javaType);
        return this;
    }

    public BodyStepMultipart receiveBytes() {
        byteArrayJavaType = objectMapper.getTypeFactory().constructParametricType(List.class, Byte.class);
        requestModel.setResponseJavaType(byteArrayJavaType);
        return this;
    }

    public FinalStep<?> and() throws Exception {
        if (requestModel.getResponseJavaType() == null) {
            return new FinalStepNoType(requestModel, requester);
        } else {
            if (requestModel.getResponseJavaType().equals(byteArrayJavaType)) {
                return new FinalStepByteType(requestModel, requester);
            } else {
                return new FinalStepJsonType<>(requestModel, requester, objectMapper);
            }
        }
    }
}

File: tools/src/main/kotlin/ru/virgil/spring_tools/tools/testing/fluent_request/BodyStepMultipart.kt
package ru.virgil.spring_tools.tools.testing.fluent_request

import com.fasterxml.jackson.databind.JavaType
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.mock.web.MockMultipartFile

class BodyStepMultipart(
    private val requestModel: RequestModel,
    private val requester: Requester,
    private val objectMapper: ObjectMapper,
) : BodyStepMultipartStart, BodyStep {

    private var byteArrayJavaType: JavaType? = null

    @Throws(Exception::class)
    override fun file(mockMultipartFile: MockMultipartFile): BodyStepMultipart {
        requestModel.mockMultipartFile = (mockMultipartFile)
        return this
    }

    fun receive(responseClass: Class<*>?, vararg responseClasses: Class<*>): BodyStepMultipart {
        val javaType = objectMapper.typeFactory.constructParametricType(responseClass, *responseClasses)
        requestModel.responseJavaType = (javaType)
        return this
    }

    fun receiveBytes(): BodyStepMultipart {
        byteArrayJavaType = objectMapper.typeFactory.constructParametricType(MutableList::class.java, Byte::class.java)
        requestModel.responseJavaType = (byteArrayJavaType)
        return this
    }

    @Throws(Exception::class)
    override fun and(): FinalStep<*> {
        return if (requestModel.responseJavaType == null) {
            FinalStepNoType(requestModel, requester)
        } else {
            if (requestModel.responseJavaType == byteArrayJavaType) {
                FinalStepByteType(requestModel, requester)
            } else {
                FinalStepJsonType<Any>(requestModel, requester, objectMapper)
            }
        }
    }
}
